062-prometheus-metrics-reporter.md

```diff
@@ -0,0 +1,202 @@
+
+At the moment Strimzi relies on these default JMX reporters and uses [jmx_exporter](https://github.com/prometheus/jmx_exporter) which is a Java agent that retrieves metrics via JMX and exposes them over an HTTP endpoint in the prometheus format. Then Prometheus is configured to scrape that endpoint to retrieve the Kafka metrics.
+
+![Current Situation](images/062-current.png)
+
+- `org.apache.kafka.common.metrics.JmxReporter` is the reporter implementation for the Kafka Metrics.
+- `org.apache.kafka.server.metrics.FilteringJmxReporter` is the reporter implementation for the Yammer Metrics. It’s named `FilteringJmxReporter` because it extends the JmxReporter class from the Yammer library and adds an option to select the metrics to report.
+
+
+## Motivation
+
+I propose updating the metrics collection pipeline for the following reasons:
+
+1. The current metrics collection pipeline is pretty convoluted. We have metrics reporters first exposing metrics via JMX before using a Java agent to expose them again via HTTP. Using metrics reporters to directly expose metrics to Prometheus would significantly simplify it by removing JMX from the picture and removing jmx_exporter.
+2. Each component along the pipeline has its own configurations and specificities. For example, since Kafka 3.4.0, it’s possible to disable the JmxReporter for KafkaMetrics by setting [auto.include.jmx.reporter](https://kafka.apache.org/documentation/#brokerconfigs_auto.include.jmx.reporter) to `false`. It is currently not possible to disable `FilteringJmxReporter` (it’s something I plan to address in the near future). Also both jmx_exporter and the built-in reporter allow selecting metrics to collect.
+3. The jmx_exporter Java agent supports complex metrics mapping rules. These rules allow renaming metrics and it can make it hard to investigate issues as metrics could have different names. This prevents using metrics like an API, whether this is to build grafana dashboards or for the operator to rely on metrics.
+4. Due to the complex rules, it performs badly when there’s a very large number of metrics due to a lot of topic/partitions.
+
+
+## Proposal
+
+The proposal is to build metrics reporters that directly exposes metrics via an HTTP endpoint in the Prometheus format. This will be a new project/repository under the Strimzi organization.
+
+![Proposal](images/062-proposal.png)
+
+Today to enable jmx_exporter, users use:
+```
+metricsConfig:
+  type: jmxPrometheusExporter
+  valueFrom:
+    configMapKeyRef:
+      name: kafka-metrics
+      key: kafka-metrics-config.yml
+```
+
+To enable metric reporters, they will instead use this for example:
+```
+metricsConfig:
+  type: strimziMetricsReporter
+  config:
+    allowList: ".*"
```

Comment:
I was wondering if we want to completely change the API related to configuration part or we want to stick with the current `configMapKeyRef` with the ConfigMap hosting what you have directly here.
From a UX perspective it would just mean having a new type but not changing the way you configure in the `Kafka` CR. Of course, the ConfigMap content would be different.

Reply:
It's hard for me to tell, I guess it's more a question for the other Strimzi maintainers.

Reply:
@strimzi/maintainers what's your take on this?

Reply:
I think it depends on how convinced we are the APi from the proposal is _final_. Having a separate config map for the `allowList` field is overkill and the currently proposed API is IMHO better. But we do not want to add new fields and move it to the config map a few months later.

Reply:
That said ... what kind of `allowList` it is? Might deserve a better name -> does it filter the metrics we provide? The topics or users for which we provide them?

Reply:
> Having a separate config map for the allowList field is overkill and the currently proposed API is IMHO better. But we do not want to add new fields and move it to the config map a few months later

Yeah I agree but how much does it cost having a ConfigMap with one field and leaving space for additions later? @mimaison any chance/idea that this config could evolve having more parameters which would make worth it to still have a ConfigMap?

Reply:
> Yeah I agree but how much does it cost having a ConfigMap with one field and leaving space for additions later?

I would say it costs a lot on _user friendliness_? You have more resources to manage, you have more logic to implement as you also need to validate things such as it pointing to the right config map etc. If you remember, we used to have the JMX Exporter configuration directly in the CR as well. The problem was that its configuration was huge unstructured (from Strimzi perspective) YAML. So we moved it out. If this sticks with a few simple fields, then having it inline is IMHO better. But there is of course the _if_ which we need to put some likelyhood on.

Reply:
The thing is ... 
You have a `type` field which, from a UX perspective, could mean that we support different types and we could add more in the future.
Then you have the configuration directly in the `config` field where some parameters could make sense for a type but not for others (i.e. allowList makes sense for this reporter but not for (maybe) future other reporters so the operator has to do some validation). It's also true at the same time that you could have the same problem in a ConfigMap where you put parameters which doesn't make sense for the reporter type you specified, but at least it's not strongly defined in the API (it's just a field in a ConfigMap).

Reply:
I'm afraid I do not understand what are you trying to say with this comment. This is pretty much the same way to configure it as we have in all other places. E.g. listeners, storage, etc. `config` typically means opaque map with configuration options. So not sure we need it here or whether we can have the allowList directly under `metricsConfig`. But the idea that each type has its own fields is something we have in many other places already.

Reply:
My bad forgetting it's an opaque map and not a strongly typed configuration section. So you are right.
Anyway, we have the history of previous reporter where we needed to move configuration out to a ConfigMap because it was getting bigger and bigger. If we are sure it's not going to happen again, we can get rid of ConfigMap. But I think it's more @mimaison who can answer this question.

Reply:
The `allowList` could get relatively long.

I used `kafka_cluster.*,kafka_controller.*,kafka_log.*,kafka_network.*,kafka_server_(brokertopicmetrics|delayedoperationpurgatory|fetchsessioncache|kafkarequesthandlerpool|kafkaserver|replicaalterlogdirsmanager|replicafetchermanager|replicamanager|sessionexpirelistener|socket_server_metrics|zookeeperclientmetrics).*,kafka_utils.*` for testing which is already not the most readable line.

Users wanting finer metrics filtering, for example only wanting request metrics for specific types of requests could easily end up with a long list of regexes.

Also I wonder if adding a `denyList` option would be useful. 

Using a ConfigMap would allow entering the patterns on different line (potentially with comments too) which is more readable IMO.

Reply:
If this is the case I would have a ConfigMap so we can stick with the current API in the CR.

Reply:
Ok so if I understand the configuration would be in this format?
```
metricsConfig:
  type: strimziMetricsReporter
  valueFrom:
    configMapKeyRef:
      name: kafka-metrics
      key: metrics-reporter-config.yml
```

Reply:
Yes something like that.

Reply:
Done